thirdparty/hnswlib/hnswlib/hnswalg.h

```diff
@@ -210,8 +228,53 @@ class HierarchicalNSW : public AlgorithmInterface<dist_t> {
     char* map_;
     size_t map_size_;
 
+    bool sq_enabled_ = false;
+    bool sq_refine_enabled_ = false;
+    float alpha_ = 0.0f;
+
     mutable knowhere::lru_cache<uint64_t, tableint> lru_cache;
 
+    void
+    trainSQuant(const float* train_data, size_t ntrain) {
+        size_t dim = *(size_t*)dist_func_param_;
+        for (size_t i = 0; i < ntrain; ++i) {
+            const float* vec = train_data + i * dim;
+            std::unique_ptr<float[]> vec_norm = nullptr;
+            if (metric_type_ == Metric::COSINE) {
+                vec_norm = knowhere::CopyAndNormalizeVecs(vec, 1, dim);
+                vec = vec_norm.get();
+            }
+            for (size_t j = 0; j < dim; ++j) {
+                alpha_ = std::max(alpha_, std::abs(vec[j]));
+            }
+        }
+    }
+
+    void
+    encodeSQuant(const float* from, int8_t* to) const {
+        size_t dim = *(size_t*)dist_func_param_;
+        std::unique_ptr<float[]> data_norm = nullptr;
+        if (metric_type_ == Metric::COSINE) {
+            data_norm = knowhere::CopyAndNormalizeVecs(from, 1, dim);
+            from = data_norm.get();
+        }
+        for (size_t i = 0; i < dim; ++i) {
```

Comment:
please confirm that is it really `int8_t`, not `uint8_t`.
Classical SQ implementation uses uint8_t and maps into [0, 255] range, not [-128, 127].

Reply:
SQ used there is symmetric version which maps to [-127, 127], which unify ip and l2 metric and simplify implementation.

Reply:
Ok